extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,165 @@
+  // This allows ads from multiple ad servers on the same page
+  if (!(e.win.hasOwnProperty('imageadBatchManagers'))) {
+    // Create an array of batch mangers. There will be one for each URL.
+    e.win.imageadBatchManagers = {};
+    // Scan the page to get a list of all image ads on this page, indexed by url.
+    // For performance reasons, we only do the scan once, the first time that a batch manager is constructed.
+    e.win.imageadElements = {};
+    const elements = document.querySelectorAll(selector);
+    for (let i = 0; i < elements.length; i++) {
+      const el = elements[i];
+      const url = el.getAttribute('data-url');
+      if (!(url in e.win.imageadElements)) {
+        e.win.imageadElements[url] = [];
+      }
+      e.win.imageadElements[url].push(el);
+    }
+  }
+  if (!(url in e.win.imageadBatchManagers)) {
+    // This is the first time we've seen this URL, so this will be the master for this batch
+    e.isBatchMaster_ = true;
+    e.win.imageadBatchManagers[url] = new AmpAdBatchManager(e);
+  } else {
+    e.isBatchMaster_ = false;
+  }
+  return e.win.imageadBatchManagers[url];
+}
+
+export class AmpAdBatchManager {
+
+  /** @param {!Element} element */
+  constructor(element) {
+
+    /** @private {string} The base URL of the ad server */
+    this.url_ = element.url_;
+
+    /** @private {string} The full URL of the ad server, including any slot ids */
+    this.fullurl_ = this.url_;
+
+    /** @private {string} The number of timer ticks remaining before we give up waiting for an ad server response */
+    this.timeout_ = AD_BATCH_TIMEOUT;
+
+    /** @public {Object} responseData Map of JSON responses indexed by slot */
+    this.responseData = null;
+
+    // Get a list of slots that are in use for this URL
+    const slots = [];
+    const elements = element.win.imageadElements[this.url_];
+    for (let i = 0; i < elements.length; i++) {
+      const e = elements[i];
+
+      // If a slot is defined, add it to the list if it isn't there already
+      let s = e.getAttribute('data-slot');
+      if (s === null) {
+        s = 0;
+      }
+      let done = false;
+      for (const x in slots) {
+        if (slots[x] == s) {
+          done = true;
+        }
+      }
+      if (!done) {
+        slots.push(s);
+      }
+    }
+    // If any slots were defined, add them to the URL
+    if (slots.length) {
+      this.fullurl_ += '?s=' + slots.join(',');
+    }
+  }
+
+  /**
+   * Call this to instruct the batch manager to do the layout for an element. Prior to doing the layout, the batch manager
+   * will fetch the data (if this is the batch master) or wait for it to be ready (if not the batch master).
+   * @param {Element} e The element for which the layout is to be done
+   * @returns {Promise} The element's "jsondata_" variable will be filled its with the data from the ad server before this is
+   *     resolved.
+   */
+  doLayout(e) {
+    const t = this;
```

Comment:
It's needed in order to be able to pass the batch manager's private and public variables to the Promise's function.  If you use "this" inside the Promise, you get an object of type Window, not AmpAdBatchManager.